src/hotspot/share/utilities/objectBitSet.inline.hpp

```diff
@@ -53,52 +52,15 @@ ObjectBitSet<F>::~ObjectBitSet() {
     delete current;
     current = next;
   }
-}
-
-template<MEMFLAGS F>
-ObjectBitSet<F>::BitMapFragmentTable::~BitMapFragmentTable() {
-  for (int index = 0; index < BasicHashtable<F>::table_size(); index ++) {
-    Entry* e = bucket(index);
-    while (e != nullptr) {
-      Entry* tmp = e;
-      e = e->next();
-      BasicHashtable<F>::free_entry(tmp);
-    }
-  }
-}
-
-template<MEMFLAGS F>
-inline typename ObjectBitSet<F>::BitMapFragmentTable::Entry* ObjectBitSet<F>::BitMapFragmentTable::bucket(int i) const {
-  return (Entry*)BasicHashtable<F>::bucket(i);
-}
-
-template<MEMFLAGS F>
-inline typename ObjectBitSet<F>::BitMapFragmentTable::Entry*
-  ObjectBitSet<F>::BitMapFragmentTable::new_entry(unsigned int hash, uintptr_t key, CHeapBitMap* value) {
-
-  Entry* entry = (Entry*)BasicHashtable<F>::new_entry(hash);
-  entry->_key = key;
-  entry->_value = value;
-  return entry;
-}
-
-template<MEMFLAGS F>
-inline void ObjectBitSet<F>::BitMapFragmentTable::add(uintptr_t key, CHeapBitMap* value) {
-  unsigned hash = hash_segment(key);
-  Entry* entry = new_entry(hash, key, value);
-  BasicHashtable<F>::add_entry(hash_to_index(hash), entry);
-}
-
-template<MEMFLAGS F>
-inline CHeapBitMap** ObjectBitSet<F>::BitMapFragmentTable::lookup(uintptr_t key) {
-  unsigned hash = hash_segment(key);
-  int index = hash_to_index(hash);
-  for (Entry* e = bucket(index); e != NULL; e = e->next()) {
-    if (e->hash() == hash && e->_key == key) {
-      return &(e->_value);
+  class Deleter {
+    public:
+    bool do_entry(uintptr_t& key, CHeapBitMap*& fragment) {
+      // fragment is already deleted above, but now delete the table nodes.
+      return true;
     }
-  }
-  return NULL;
+  };
+  Deleter deleter;
+  _bitmap_fragments.unlink(&deleter);
```

Comment:
This shouldn't be necessary. The destructor of _bitmap_fragments will be implicitly called and will free all of its entries.

Reply:
Yes, I added a temporary assert and indeed it is.  Thanks for pointing this out.

Reply:
I'd actually tested with the version without it and then thought I needed it this morning, so I'm glad I don't need it.